source/sigv4.c

```diff
@@ -436,6 +561,402 @@ static SigV4Status_t parseDate( const char * pDate,
+        returnVal = SigV4Success;
+    }
+
+    return returnVal;
+}
+
+/*-----------------------------------------------------------*/
+
+#if ( SIGV4_USE_CANONICAL_SUPPORT == 1 )
+
+    static void stringToSigV4Value( SigV4String_t * pSigV4Value,
+                                    const char * pInput,
+                                    size_t lenInput )
+    {
+        assert( pSigV4Value != NULL );
+        assert( pInput != NULL );
+        assert( lenInput > 0U );
+
+        pSigV4Value->pData = ( char * ) pInput;
+        pSigV4Value->dataLen = ( size_t ) lenInput;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static bool emptySigV4String( SigV4String_t * pInput )
+    {
+        bool returnVal = true;
+
+        assert( pInput != NULL );
+
+        return ( pInput->pData == NULL || pInput->dataLen == 0 ) ? returnVal : !returnVal;
+    }
+
+/*-----------------------------------------------------------*/
+
+    static int cmpKeyValue( const void * pFirstVal,
+                            const void * pSecondVal )
+    {
+        SigV4KeyValuePair_t * pFirst, * pSecond = NULL;
+        size_t lenSmall = 0U;
+
+        assert( pFirstVal != NULL );
+        assert( pSecondVal != NULL );
+
+        pFirst = ( SigV4KeyValuePair_t * ) pFirstVal;
+        pSecond = ( SigV4KeyValuePair_t * ) pSecondVal;
+
+        assert( !emptySigV4String( &pFirst->key ) );
+        assert( !emptySigV4String( &pSecond->key ) );
+
+        if( pFirst->key.dataLen <= pSecond->key.dataLen )
+        {
+            lenSmall = pFirst->key.dataLen;
+        }
+        else
+        {
+            lenSmall = pSecond->key.dataLen;
+        }
+
+        return strncmp( ( char * ) pFirst->key.pData,
+                        ( char * ) pSecond->key.pData,
+                        lenSmall );
+    }
+
+/*-----------------------------------------------------------*/
+
+    static void encodeURI( const char * pURI,
+                           size_t uriLen,
+                           char * pCanonicalURI,
+                           size_t * canonicalURILen,
+                           bool encodeSlash,
+                           bool nullTerminate )
+    {
+        const char * pURILoc = NULL;
+        char * pBufLoc = NULL;
+        size_t index = 0U;
+
+        assert( pURI != NULL );
+        assert( pCanonicalURI != NULL );
+        assert( canonicalURILen != NULL );
+        assert( *canonicalURILen > 0U );
+
+        pURILoc = ( const char * ) pURI;
+        pBufLoc = ( char * ) pCanonicalURI;
+
+        while( index < uriLen && *pURILoc )
+        {
+            if( isalnum( *pURILoc ) || ( *pURILoc == '-' ) || ( *pURILoc == '_' ) || ( *pURILoc == '.' ) || ( *pURILoc == '~' ) )
+            {
+                *pBufLoc++ = *pURILoc;
+                index++;
+            }
+            else if( ( *pURILoc == '/' ) && !encodeSlash )
+            {
+                *pBufLoc++ = *pURILoc;
```

Comment:
Please use brackets on this post increment to make it clear if we are intending to increment the pointer or the value it is pointing to.